modules/auxiliary/client/telegram/send_message.rb

```diff
@@ -3,54 +3,104 @@
-        OptString.new('MSG', [true, 'Message content', 'New session opened!']),
-        OptEnum.new('FORMATTING', [true, 'Message formating option (Markdown|MarkdownV2|HTML)', 'Markdown', [ 'Markdown', 'MarkdownV2', 'HTML']]),
+        OptString.new('MESSAGE', [false, 'Optional message sent with the document']),
+        OptInt.new('CHAT_ID', [false, 'Chat ID for the BOT', '']),
+        OptPath.new('DOCUMENT', [false, 'The path to the document(binary, video etc)']),
+        OptPath.new('IDFILE', [false, 'File containing chat IDs, one per line'])
       ], self.class
     )
   end
 
   def message
-    datastore['MSG']
+    datastore['MESSAGE']
   end
 
-  def formatting
-    datastore['FORMATTING']
+  def document
+    datastore['DOCUMENT']
   end
 
   def bot_token
     datastore['BOT_TOKEN']
   end
 
-  def run
-    uri = URI("https://api.telegram.org/bot#{bot_token}/sendMessage")
-    params = { chat_id: datastore['CHAT_ID'], parse_mode: formatting, text: message }
-    uri.query = URI.encode_www_form(params)
-    res = Net::HTTP.get_response(uri)
+  def id_file
+    datastore['IDFILE']
+  end
 
-    if res.is_a?(Net::HTTPSuccess)
-      print_good('Message sent!')
+  def send_document(conn, chat_id)
+    return unless document
+
+    raw_params = { 'chat_id' => chat_id, 'document' => Faraday::UploadIO.new(document, 'application/octet-stream') }
+    parms = {}
+    raw_params.each_with_object({}) do |(key, value), _params|
+      parms[key] = value
+    end
+    print_warning("Sending to #{chat_id.strip}")
+    response = conn.post("/bot#{bot_token}/sendDocument", parms)
+    if response.status == 200
+      print_good('Document sent successfully!')
+    elsif response.status == 403
+      print_bad("Error while sending document! Make sure the user with chat_id : #{chat_id} is active on bot.")
     else
-      print_error('Unable to send the message!')
-      print_error("API Status: #{res.code}")
+      print_bad('Error while sending the document!')
     end
   end
+
+  def send_message(conn, chat_id)
+    return if message == ''
+
+    parms = { 'chat_id' => chat_id, 'text' => message }
+    response = conn.post("/bot#{bot_token}/sendMessage", parms)
+    if response.status == 200
+      print_good('Message sent successfully')
+    elsif response.status == 403
+      print_bad("Error while sending messsage. Make sure the user with chat_id : #{chat_id} is active on bot")
+    else
+      print_bad('Error while sending the message.')
+    end
+  end
+
+  def run
+    url = 'https://api.telegram.org'
+    conn = Faraday.new(url: url) do |faraday|
+      faraday.request :multipart
+      faraday.request :url_encoded
+      faraday.adapter Faraday.default_adapter
+    end
+
+    if id_file
+      File.readlines(id_file).each do |chat_id|
+        send_document(conn, chat_id)
+        send_message(conn, chat_id)
+      end
+      return
+    end
+    send_document(conn, datastore['CHAT_ID'])
+    send_message(conn, datastore['CHAT_ID'])
```

Comment:
For this and the code above, what if I only want to send a message or only want to send a document? The ability to send a document should be optional and not required.

Reply:
Yes, both message and document are optional, a user can send either of them or both. But I'll add a check to see if user enters atleast one of them :+1:

Reply:
Actually both the methods have `return unless message/document` statement which will check if the message/document is there, if not it will return without continuing further. So in case the user has just set the `message`, it will only send the message.

Reply:
Hmm so I think you have the right idea here actually, and I'm going to close some of my previous advice down now that I've reviewed the code a bit more. Basically I think the main thing that needs to be done here is to say "if `document` and `message` are both not provided, then print an error message as we need one or the other".

Reply:
This was resolved with https://github.com/rapid7/metasploit-framework/pull/15178/commits/03d66ac4bc4e6f5905cb10cf6ec5ce8c57e0d8ff